### PR TITLE
Extend MMULT to support non-square matrices

### PR DESCRIPTION
**Pull request by dr-draik**
The current implementation of MMULT only supports square matrices. This extends to support non-square matrices.

Matrix multiplication only requires that the number of columns in the first matrix are the same as the number of rows in the second matrix. The resulting matrix has dimensions `matrix1Rows x matrix2Cols`.

### CODE REVIEW

**Comment by igitur** [comment]
Thanks, I'll look at this soon.

**Review by Pankraty — APPROVED** [review]
Looks good for me!

**Comment by igitur** [comment]
I just still want to check whether all the edge cases are handled (eg errors) and whether there are unit tests for them all.

File: ClosedXML.Tests/Excel/CalcEngine/FunctionsTests.cs [hunk 1]
```diff
@@ -249,6 +249,57 @@ public void MMult()
...
+
+            ws.Cell("A1").FormulaA1 = $"MMULT({array1Range},{array2Range})";
+
+            var error = Assert.Throws<ArgumentException>(() =>
+                {
+                    var val = ws.Cell("A1").Value;
+                }
+            );
+
+            Assert.AreEqual("Range 1 must have the same number of columns as range 2 has rows.", error.Message);
```
Comment (igitur) [inline ClosedXML.Tests/Excel/CalcEngine/FunctionsTests.cs:300]: Please reword to `The number of columns in array1 is different from the number of rows in array2.`

File: ClosedXML.Tests/Excel/CalcEngine/FunctionsTests.cs [hunk 2]
```diff
@@ -249,6 +249,57 @@ public void MMult()
...
+        public void MMult_ThrowsWhenArray1RowsNotEqualToArray2Cols(string array1Range, string array2Range)
+        {
+            IXLWorksheet ws = new XLWorkbook().AddWorksheet("Sheet1");
+
+            ws.Cells($"{array1Range}").Value = 1.0;
+            ws.Cells($"{array2Range}").Value = 1.0;
+
+            ws.Cell("A1").FormulaA1 = $"MMULT({array1Range},{array2Range})";
+
+            var error = Assert.Throws<ArgumentException>(() =>
```
Comment (igitur) [inline ClosedXML.Tests/Excel/CalcEngine/FunctionsTests.cs:294]: Please refer to https://support.microsoft.com/en-us/office/mmult-function-40593ed7-a3cd-4b6b-b9a3-e4ad3c7245eb?ns=excel&version=19&syslcid=1033&uilcid=1033&appver=zxl190&helpid=xlmain11.chm60213&ui=en-us&rs=en-us&ad=us

The correct exception to be thrown is `CellValueException`. I see this is a legacy issue, but please add a commit to fix this.

Bonus points if you can also add a test for when any cells are empty or contain text.

**Review by igitur — CHANGES_REQUESTED** [review]

**Comment by igitur** [comment]
Ref:
image

File: ClosedXML.Tests/Excel/CalcEngine/FunctionsTests.cs (hunk 2 above)
Comment (igitur) [reply ClosedXML.Tests/Excel/CalcEngine/FunctionsTests.cs:294]: Yes, please. That sounds perfect.

Comment (dr-draik) [reply ClosedXML.Tests/Excel/CalcEngine/FunctionsTests.cs]: I've switched the `ArgumentException` for size mismatches to `CellValueException`.

I've also added the `CellValueException` wrapper for format exceptions emerging from `MathTrig.GetArray`, and a test covering a few cases.

File: ClosedXML.Tests/Excel/CalcEngine/FunctionsTests.cs (hunk 1 above)
Comment (dr-draik) [reply ClosedXML.Tests/Excel/CalcEngine/FunctionsTests.cs]: Updated accordingly.

**Comment by igitur** [comment]
Oh, I see a fellow Capetownian! Welcome.

**Comment by dr-draik** [comment]
Thanks! Yeah, I'm currently building up my winter resistance - I always forget how frigid it ends up getting here.

**Review by igitur — APPROVED** [review]

**Comment by Pankraty** [comment]
Cape Town in winter:
image

Saratov in summer:
image

😜

**Comment by igitur** [comment]
> Cape Town in winter:
> 
> Saratov in summer:
> 😜

That's why I was tasked to wash the bedding *today*!
